// @flood/element/src/page/locators/LinkText.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Locator_1 = require("../Locator");
class LinkTextLocator extends Locator_1.BaseLocator {
    constructor(linkText, partial = false, desc) {
        super(desc);
        this.linkText = linkText;
        this.partial = partial;
    }
    get pageFuncArgs() {
        return [this.linkText, this.partial];
    }
    get pageFunc() {
        return (targetText, partial) => {
            let links = Array.from(document.querySelectorAll('body a'));
            return links.find(link => {
                if (!link.textContent)
                    return false;
                let text = link.textContent.trim();
                return (partial && text.indexOf(targetText) !== -1) || text === targetText;
            });
        };
    }
    get pageFuncMany() {
        return (targetText, partial) => {
            let links = Array.from(document.querySelectorAll('body a'));
            return links.filter(link => {
                if (!link.textContent)
                    return false;
                let text = link.textContent.trim();
                return (partial && text.indexOf(targetText) !== -1) || text === targetText;
            });
        };
    }
    toString() {
        if (this.partial) {
            return `partial link text "${this.linkText}"`;
        }
        else {
            return `link text "${this.linkText}"`;
        }
    }
}
exports.LinkTextLocator = LinkTextLocator;
