// @flood/element/src/page/Enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MouseButtons;
(function (MouseButtons) {
    MouseButtons["LEFT"] = "left";
    MouseButtons["RIGHT"] = "right";
    MouseButtons["MIDDLE"] = "middle";
})(MouseButtons = exports.MouseButtons || (exports.MouseButtons = {}));
/**
 * Chrome DevTools Device Emulation
 */
var Device;
(function (Device) {
    Device["blackberryPlayBook"] = "Blackberry PlayBook";
    Device["blackberryPlayBookLandscape"] = "Blackberry PlayBook landscape";
    Device["blackBerryZ30"] = "BlackBerry Z30";
    Device["blackBerryZ30Landscape"] = "BlackBerry Z30 landscape";
    Device["galaxyNote_3"] = "Galaxy Note 3";
    Device["galaxyNote_3Landscape"] = "Galaxy Note 3 landscape";
    Device["galaxyNoteIi"] = "Galaxy Note II";
    Device["galaxyNoteIiLandscape"] = "Galaxy Note II landscape";
    Device["galaxySIii"] = "Galaxy S III";
    Device["galaxySIiiLandscape"] = "Galaxy S III landscape";
    Device["galaxyS5"] = "Galaxy S5";
    Device["galaxyS5Landscape"] = "Galaxy S5 landscape";
    Device["iPad"] = "iPad";
    Device["iPadLandscape"] = "iPad landscape";
    Device["iPadMini"] = "iPad Mini";
    Device["iPadMiniLandscape"] = "iPad Mini landscape";
    Device["iPadPro"] = "iPad Pro";
    Device["iPadProLandscape"] = "iPad Pro landscape";
    Device["iPhone4"] = "iPhone 4";
    Device["iPhone4Landscape"] = "iPhone 4 landscape";
    Device["iPhone5"] = "iPhone 5";
    Device["iPhone5Landscape"] = "iPhone 5 landscape";
    Device["iPhone6"] = "iPhone 6";
    Device["iPhone6Landscape"] = "iPhone 6 landscape";
    Device["iPhone6Plus"] = "iPhone 6 Plus";
    Device["iPhone6PlusLandscape"] = "iPhone 6 Plus landscape";
    Device["iPhoneX"] = "iPhone X";
    Device["iPhoneXLandscape"] = "iPhone X landscape";
    Device["kindleFireHdx"] = "Kindle Fire HDX";
    Device["kindleFireHdxLandscape"] = "Kindle Fire HDX landscape";
    Device["lgOptimusL70"] = "LG Optimus L70";
    Device["lgOptimusL70Landscape"] = "LG Optimus L70 landscape";
    Device["microsoftLumia550"] = "Microsoft Lumia 550";
    Device["microsoftLumia950"] = "Microsoft Lumia 950";
    Device["microsoftLumia950Landscape"] = "Microsoft Lumia 950 landscape";
    Device["nexus10"] = "Nexus 10";
    Device["nexus10Landscape"] = "Nexus 10 landscape";
    Device["nexus4"] = "Nexus 4";
    Device["nexus4Landscape"] = "Nexus 4 landscape";
    Device["nexus5"] = "Nexus 5";
    Device["nexus5Landscape"] = "Nexus 5 landscape";
    Device["nexus5X"] = "Nexus 5X";
    Device["nexus5XLandscape"] = "Nexus 5X landscape";
    Device["nexus6"] = "Nexus 6";
    Device["nexus6Landscape"] = "Nexus 6 landscape";
    Device["nexus6P"] = "Nexus 6P";
    Device["nexus6PLandscape"] = "Nexus 6P landscape";
    Device["nexus7"] = "Nexus 7";
    Device["nexus7Landscape"] = "Nexus 7 landscape";
    Device["nokiaLumia_520"] = "Nokia Lumia 520";
    Device["nokiaLumia_520Landscape"] = "Nokia Lumia 520 landscape";
    Device["nokiaN9"] = "Nokia N9";
    Device["nokiaN9Landscape"] = "Nokia N9 landscape";
})(Device = exports.Device || (exports.Device = {}));
/**
 * Lists all available keyboard control keys which can be used when sending a key press combination.
 *
 * @export
 * @enum {number}
 */
exports.Key = {
    NULL: null,
    CANCEL: 'Cancel',
    HELP: 'Help',
    BACK_SPACE: 'Backspace',
    TAB: 'Tab',
    CLEAR: 'Clear',
    RETURN: 'Enter',
    ENTER: 'Enter',
    SHIFT: 'ShiftLeft',
    CONTROL: 'ControlLeft',
    ALT: 'AltLeft',
    PAUSE: 'Pause',
    ESCAPE: 'Escape',
    SPACE: 'Space',
    PAGE_UP: 'PageUp',
    PAGE_DOWN: 'PageDown',
    END: 'End',
    HOME: 'Home',
    ARROW_LEFT: 'ArrowLeft',
    LEFT: 'ArrowLeft',
    ARROW_UP: 'ArrowUp',
    UP: 'ArrowUp',
    ARROW_RIGHT: 'ArrowRight',
    RIGHT: 'ArrowRight',
    ARROW_DOWN: 'ArrowDown',
    DOWN: 'ArrowDown',
    INSERT: 'Insert',
    DELETE: 'Delete',
    SEMICOLON: 'Semicolon',
    EQUALS: 'Equal',
    NUMPAD0: 'Numpad0',
    NUMPAD1: 'Numpad1',
    NUMPAD2: 'Numpad2',
    NUMPAD3: 'Numpad3',
    NUMPAD4: 'Numpad4',
    NUMPAD5: 'Numpad5',
    NUMPAD6: 'Numpad6',
    NUMPAD7: 'Numpad7',
    NUMPAD8: 'Numpad8',
    NUMPAD9: 'Numpad9',
    MULTIPLY: 'NumpadMultiply',
    ADD: 'NumpadAdd',
    SEPARATOR: null,
    SUBTRACT: 'NumpadSubtract',
    DECIMAL: 'NumpadDecimal',
    DIVIDE: 'NumpadDivide',
    F1: 'F1',
    F2: 'F2',
    F3: 'F3',
    F4: 'F4',
    F5: 'F5',
    F6: 'F6',
    F7: 'F7',
    F8: 'F8',
    F9: 'F9',
    F10: 'F10',
    F11: 'F11',
    F12: 'F12',
    COMMAND: 'MetaLeft',
    META: 'MetaLeft',
};
exports.userAgents = {
    iphone: 'Mozilla/5.0 (iPhone; CPU iPhone OS 5_1 like Mac OS X) AppleWebKit/534.46 (KHTML, like Gecko) Version/5.1 Mobile/9B179 Safari/7534.48.3',
    ipod: 'Mozilla/5.0 (iPod; CPU iPhone OS 5_1 like Mac OS X) AppleWebKit/534.46 (KHTML, like Gecko) Version/5.1 Mobile/9B176 Safari/7534.48.3',
    ipad: 'Mozilla/5.0 (iPad; CPU OS 5_1 like Mac OS X) AppleWebKit/534.46 (KHTML, like Gecko) Version/5.1 Mobile/9B176 Safari/7534.48.3',
    safariOSX: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_4) AppleWebKit/536.25 (KHTML, like Gecko) Version/6.0 Safari/536.25',
    safariWindows: 'Mozilla/5.0 (Windows; Windows NT 6.1) AppleWebKit/534.57.2 (KHTML, like Gecko) Version/5.1.7 Safari/534.57.2',
    ie7: 'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.0)',
    ie8: 'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.0; Trident/4.0)',
    ie9: 'Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)',
    chromeOSX: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_4) AppleWebKit/536.5 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36',
    chromeWindows: 'Mozilla/5.0 (Windows; Windows NT 6.1) AppleWebKit/536.5 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36',
    ffOSX: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.7; rv:11.0) Gecko/20100101 Firefox/11.0',
    ffWindows: 'Mozilla/5.0 (Windows NT 6.1; rv:11.0) Gecko/20100101 Firefox/11.0',
    operaOSX: 'Opera/9.80 (Macintosh; Intel Mac OS X 10.7.4; U; en) Presto/2.10.229 Version/11.62',
    operaWindows: 'Opera/9.80 (Windows NT 6.1; U; en) Presto/2.10.229 Version/11.62',
    robot: 'I AM ROBOT',
};
