// @flood/element/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// CLI / runtime
var Reporter_1 = require("./src/Reporter");
exports.TestEvent = Reporter_1.TestEvent;
// XYZ
var Element_1 = require("./src/Element");
exports.runCommandLine = Element_1.runCommandLine;
exports.runUntilExit = Element_1.runUntilExit;
const WorkRoot_1 = require("./src/runtime-environment/WorkRoot");
exports.WorkRoot = WorkRoot_1.default;
var NullRuntimeEnvironment_1 = require("./src/runtime-environment/NullRuntimeEnvironment");
exports.nullRuntimeEnvironment = NullRuntimeEnvironment_1.nullRuntimeEnvironment;
const Puppeteer_1 = require("./src/driver/Puppeteer");
exports.launch = Puppeteer_1.launch;
exports.PuppeteerClient = Puppeteer_1.PuppeteerClient;
var TestScript_1 = require("./src/TestScript");
exports.TestScriptError = TestScript_1.TestScriptError;
var EvaluatedScript_1 = require("./src/runtime/EvaluatedScript");
exports.EvaluatedScript = EvaluatedScript_1.EvaluatedScript;
const Expect_1 = require("./src/utils/Expect");
exports.expect = Expect_1.expect;
// Test observer integration
const Test_1 = require("./src/runtime/Test");
exports.Test = Test_1.default;
var Observer_1 = require("./src/runtime/test-observers/Observer");
exports.NoOpTestObserver = Observer_1.NoOpTestObserver;
var Types_1 = require("./src/runtime/errors/Types");
exports.castStructuredError = Types_1.castStructuredError;
var ObjectTrace_1 = require("./src/utils/ObjectTrace");
exports.NullObjectTrace = ObjectTrace_1.NullObjectTrace;
var StructuredError_1 = require("./src/utils/StructuredError");
exports.StructuredError = StructuredError_1.StructuredError;
const Recorder_1 = require("./src/network/Recorder");
exports.NetworkRecorder = Recorder_1.default;
var Protocol_1 = require("./src/network/Protocol");
exports.Entry = Protocol_1.Entry;
const Observer_2 = require("./src/runtime/Observer");
exports.NetworkObserver = Observer_2.default;
