// @flood/element/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @docPage Until
 */
var Until_1 = require("./src/page/Until");
exports.Until = Until_1.Until;
/**
 * @docPage Until
 */
var Condition_1 = require("./src/page/Condition");
exports.Condition = Condition_1.Condition;
/**
 * @docPage Constants
 */
var Enums_1 = require("./src/page/Enums");
exports.Device = Enums_1.Device;
exports.Key = Enums_1.Key;
exports.MouseButtons = Enums_1.MouseButtons;
/**
 * @docPage By
 */
var By_1 = require("./src/page/By");
exports.By = By_1.By;
/**
 * @docPage TestData
 * @docAlias TestDataFactory TestData
 */
var TestData_1 = require("./src/test-data/TestData");
exports.TestDataSource = TestData_1.TestDataSource;
const TestDataLoaders_1 = require("./src/test-data/TestDataLoaders");
/**
 * `TestData` is a pre-configured instance of <[TestDataFactory]> that can be used to prepare test data for your script.
 *
 * **Example**
 * ```typescript
 * import { step, Browser, TestData, TestSettings } from '@flood/element'
 *
 * interface Row {
 *   username: string
 *   userID: number
 * }
 * TestData.fromCSV<Row>('users.csv').shuffle()
 * ```
 *
 * @docPage TestData
 */
exports.TestData = new TestDataLoaders_1.NullTestDataLoaders();
/**
 * @docPage Settings
 */
var Settings_1 = require("./src/runtime/Settings");
exports.setup = Settings_1.setup;
exports.DEFAULT_SETTINGS = Settings_1.DEFAULT_SETTINGS;
const types_1 = require("./src/runtime-environment/types");
/**
 * A subset of `process.env` available to this test. It is of type <[FloodProcessEnv]>.
 * @docPage DSL
 */
exports.ENV = types_1.nullFloodProcessEnv;
/**
 * @docPage DSL
 */
var Step_1 = require("./src/runtime/Step");
exports.step = Step_1.step;
/**
 * @docPage DSL
 */
var types_2 = require("./src/runtime/types");
exports.suite = types_2.suite;
