// @flood/element/src/Reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// NOTE: these need to track the Lifecycle enum in control.proto
var TestEvent;
(function (TestEvent) {
    TestEvent[TestEvent["Setup"] = 1] = "Setup";
    TestEvent[TestEvent["BeforeTest"] = 2] = "BeforeTest";
    TestEvent[TestEvent["BeforeStep"] = 3] = "BeforeStep";
    TestEvent[TestEvent["BeforeStepAction"] = 4] = "BeforeStepAction";
    TestEvent[TestEvent["AfterStepAction"] = 5] = "AfterStepAction";
    TestEvent[TestEvent["StepSucceeded"] = 6] = "StepSucceeded";
    TestEvent[TestEvent["StepFailed"] = 7] = "StepFailed";
    TestEvent[TestEvent["StepSkipped"] = 8] = "StepSkipped";
    TestEvent[TestEvent["AfterStep"] = 9] = "AfterStep";
    TestEvent[TestEvent["TestSucceeded"] = 10] = "TestSucceeded";
    TestEvent[TestEvent["TestFailed"] = 11] = "TestFailed";
    TestEvent[TestEvent["AfterTest"] = 12] = "AfterTest";
    TestEvent[TestEvent["Exit"] = 13] = "Exit";
})(TestEvent = exports.TestEvent || (exports.TestEvent = {}));
